### PR TITLE
 Rails Capstone Project Expense Tracker

### PR DESCRIPTION
**Pull request by yosaddis**
# In this capstone project I have completed the following features

Demo of the project [here](https://www.loom.com/share/7b5a516fe9784a64ae828aaa9c551098?t=263&sid=19bc16a8-eb51-4509-9c75-6bfe8182b6e2)
The deployed version of the project [here](https://rails-capstone-65y7.onrender.com/l)

Splash screen

A simple page with the name of your app (yes, you need to choose one), and links to the sign-up and login pages.
Sign-up and login pages

The user should be able to register in the app with full name, email, and password (all mandatory).
The user can log into the app using email and password.
If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).
Home page (categories page)

When the user logs in, they are presented with the categories page.
For each category, the user can see their name, icon, and the total amount of all the transactions that belong to that category.
When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
There is a button "add a new category" at the bottom that brings the user to the page to create a new category.
Transactions page

For a given category, the list of transactions is presented, ordered by the most recent.
At the top of the page, the user can see the total amount for the category (the sum of all of the amounts of the transactions in that category).
There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
When the user clicks on the "Back" button (<), the user navigates to the home page.
"Add a new category" page

The user fills out a form to create a new category, indicating their name and icon (both mandatory).
The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
When the user clicks on the "Back" button (<), the user navigates to the home page.
"Add a new transaction" page

The user fills out a form to create a new transaction with:
name (mandatory)
amount (mandatory)
categories (mandatory at least one)
The user clicks (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.
Unit Tests for the models and views

Thank you for your time.

### CODE REVIEW

File: app/views/category/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    <% @categories.each do |category| %>
+      <%= link_to category_expense_index_path category do %>
+        <%= render category %>
+      <% end %>
+    <% end %>
+  <% end %>
+</div>
+
+<div class="new_category_btn text_white">
+  <%= link_to "Add Category", new_category_path, class: "fcc-btn" %>
```
Comment (Lucas-Masaba) [inline app/views/category/index.html.erb:21]: - [ ] Please make sure this button appears at the bottom of the page and looks like that provided in the original design. It should have a white border surrounding the button text. According to the design guidelines, you should follow:
 - Colors
 - Typography: font face, size and weight.
 - Layout: composition and space between elements.

### Your implementation
image
### Required
image

File: app/views/expense/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      <% @expenses.each do |expense| %>
+        <%= render expense %>
+      <% end %>
+    <% end %>
+  </div>
+
+  <div class="total_txt text-center">TOTAL: $<%=@category.total_amount%> </div>
+
+  <div class="new_category_btn">
+    <%= link_to "New expense", new_category_expense_path(@category), class: "btn btn-primary add_categ_btn" %>
```
Comment (Lucas-Masaba) [inline app/views/expense/index.html.erb]: - [ ] Please make sure this button appears at the bottom of the page and looks like that provided in the original design. It should have a white border surrounding the button text. According to the design guidelines, you should follow:
 - Colors
 - Typography: font face, size and weight.
 - Layout: composition and space between elements.
 
**NB:** Please follow the same steps I mentioned for the category page.

File: app/views/expense/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    <% if @expenses.length == 0 %>
+      <p class="empty_data text-center">No expense yet. Add a new one!!!</p>
+    <% else %>
+      <% @expenses.each do |expense| %>
+        <%= render expense %>
+      <% end %>
+    <% end %>
+  </div>
+
+  <div class="total_txt text-center">TOTAL: $<%=@category.total_amount%> </div>
```
Comment (Lucas-Masaba) [inline app/views/expense/index.html.erb]: - [ ] Your total should be on the top of the page and not at the bottom as seen in the original design layout.
### Your Implementation
image
### Required
image

**Review by Lucas-Masaba — CHANGES_REQUESTED** [review]
# Hello @yosaddis  👋 , 

Good job so far! 👍 
There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there

## Highlights 🍾 
- Linter checks passing :heavy_check_mark: 
- Met most of the project requirements :gear: 
- Good use of gitflow :surfer: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions 💭 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Lucas-Masaba in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/category/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    <% @categories.each do |category| %>
+      <%= link_to category_expense_index_path category do %>
+        <%= render category %>
+      <% end %>
+    <% end %>
+  <% end %>
+</div>
+
+<div class="new_category_btn">
+  <%= link_to "Add Category", new_category_path, class: "fcc-btn" %>
```
Comment (cynthiainga) [inline app/views/category/index.html.erb:21]: - [ ] As mentioned by the previous reviewer kindly follow the given design 🙃 

Your design

Screenshot 2023-09-16 at 19 27 39

Given design

Screenshot 2023-09-16 at 19 33 42

File: app/views/expense/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    <% else %>
+      <% @expenses.each do |expense| %>
+        <%= render expense %>
+      <% end %>
+    <% end %>
+  </div>
+
+ 
+  <div class="new_category_btn">
+    <%= link_to "New expense", new_category_expense_path(@category), class: "fcc-btn" %>
```
Comment (cynthiainga) [inline app/views/expense/index.html.erb:29]: - [ ] As mentioned by the previous reviewer kindly follow the given design 🙃 

Your design

Screenshot 2023-09-16 at 19 27 49

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @yosaddis, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Highlights

- [ ] There are Home and Transaction pages ✔️ 
- [ ] User can add new transaction and new category ✔️ 
- [ ] Use devise for authentication ✔️ 
- [ ] Use Postgres as the database for the application ✔️ 

 ## Required Changes ♻️

- [ ] Please kindly check the comments under the review 👇🏾 👇🏾 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KanzaTahreem — APPROVED** [review]
# Approved 🙌

Hi @yosaddis 

Good job implementing the suggested changes 💪 **_Your project is COMPLETE! There is nothing else to say other than... it's time to merge it! 🚀 Congratulations!_**



## Highlights 🎯
There are some excellent points that you have made perfectly. Keep up the great work! 👏

-  ✅ Styling is looking great
-  ✅ User can add a new transaction
-  ✅ User can add a new category 
-  ✅ Linters are working properly

## Optional suggestions ❕
N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**